### PR TITLE
Port SlippiUser to Rust layer

### PR DESCRIPTION
**Pull request by ryanmcgrath**
Initial `SlippiUser` and `UserInfo` scaffolding. Opening this as a draft to track the effort, so nothing needed yet.

The end goal will be that `SlippiUser` in C++ just wraps this for now to reduce any churn in the Ishiiruka/mainline moves.

### CODE REVIEW

**Comment by ryanmcgrath** [comment]
Alright, this should be ready to look over now. Some high level notes:

`SlippiUser` was migrated into Rust as a module that exposes a type named `UserManager`. This works similar to the C++ version, managing login routines and a background thread operation for watching and parsing `user.json` files. `UserManager` is a thread-safe type that can be cloned into other modules and used to grab information about the current user.

Since the user is accessed on multiple threads, it's held behind an `Arc<Mutex<>>`. I threw together two small helper methods for accessing info without dealing with the locking every time - e.g:

``` rust
use slippi_user::UserManager;

fn my_operation(user_manager: UserManager) {
    let play_key = user_manager.get(|user| {
        user.play_key.clone()
    });
}
```

I've tested the new `UserManager` in most of the scenarios I could think of, but I of course welcome other testing since there's like some stuff I missed.

Some other noteworthy changes in this PR:

- `slprs_exi_device_create` now accepts a configuration struct, which just makes it easier at a glance to see exactly what's being passed over to the Rust side as parameters get added.
- `slippi_semver_str` is persisted to the Rust side, used to keep the `-beta` check in the user server overwrite call.
- `SlippiGameReporter` was renamed to `GameReporter` to stay in line with the naming conventions of `Jukebox` and `UserManager`.
- `GameReporter` now uses the new `UserManager`.
- Chat message options are now stored on the Rust side and returned over the FFI boundary as needed.
- There are a number of extra FFI handles to enable `SlippiUser` on the C++ side to keep its API, in the interest of the least churn over there while we slowly move things. These will go away in time.
- I added an additional helper method in the `ffi` root for shuffling in and out of pointer instances, just to reduce verbosity. The safety levels of these operations are unchanged.
- Different modules now accept the `http_client` as a parameter so they all share a client that's backed by the same connection pool.

If there's any questions, lemme know I can add more context. Otherwise this should be ready to test!

File: user/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,318 @@
...
+
+            return false;
+        },
+    }
+}
+
+/// The core payload that represents user information. This type is expected to conform
+/// to the same definition that the remote server uses.
+#[derive(Debug, Default, serde::Deserialize)]
+pub struct APIResponse {
```
Comment (NikhilNarayana) [inline user/src/lib.rs:258]: does this need to be public? it doesn't seem like it, but if it does then i'd like a better name so you can tell at a glance what you're getting.
maybe something like `GetUserInfoResp`

File: user/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,318 @@
...
+
+// use dolphin_integrations::Log;
+
+mod chat;
+pub use chat::DEFAULT_CHAT_MESSAGES;
+
+mod watcher;
+use watcher::UserInfoWatcher;
+
+const USER_API_URL: &'static str = "https://users-rest-dot-slippi.uc.r.appspot.com/user";
```
Comment (NikhilNarayana) [inline user/src/lib.rs:17]: add a todo for swapping over to gql

**Review by NikhilNarayana — CHANGES_REQUESTED** [review]
small nit picks, overall looks good

File: user/src/lib.rs (hunk 1 above)
Comment (ryanmcgrath) [reply user/src/lib.rs:258]: Ah, it does _not_ need to be public, no - good catch.

I think this stems from originally having the API call in a separate module while I was mocking it out. Easy swap tho.

File: user/src/lib.rs (hunk 2 above)
Comment (ryanmcgrath) [reply user/src/lib.rs:17]: Aye aye.

File: user/src/lib.rs (hunk 1 above)
Comment (ryanmcgrath) [reply user/src/lib.rs]: Done (renamed for clarity and removed `pub`).

File: user/src/lib.rs (hunk 2 above)
Comment (ryanmcgrath) [reply user/src/lib.rs:17]: Done.

**Comment by ryanmcgrath** [comment]
Comments addressed.

One thing that we should maybe take the time to do in this PR real quick is centralize the Rust logging a bit? I feel like the Jukebox should keep its own logs since audio stuff is often more log-heavy, but Fizzi had mentioned maybe condensing the rest down into either a `Slippi Online (Rust)` or just into `Slippi Online` by default.

I tend to do logs per module since it's just the norm for how I'd ship libs, buuuut this is a dylib masquerading as an app so things are def a bit different. Any thoughts?

If we don't want to change anything then I need to add a `Slippi User (Rust)` or something for now before this can be merged.

**Comment by NikhilNarayana** [comment]
i agree that we should do some log consolidation, but bundling everything (including future changes) into a single log type or into a log type that includes c++ logs doesn't seem ideal

i think we can start by consolidating everything under `Slippi Online (Rust)` for now since we don't have much non-online functionality in rust land and then consider how we want to handle unrelated logs in the future

**Review by NikhilNarayana — DISMISSED** [review]

File: dolphin/src/logger/mod.rs [hunk 3]
```diff
@@ -44,13 +44,10 @@ pub mod Log {
...
     /// to fix your logging calls. :)
-    pub const DEPENDENCIES: &'static str = "SLIPPI_RUST_DEPENDENCIES";
+    pub const Dependencies: &'static str = "SLIPPI_RUST_DEPENDENCIES";
 
     /// The default target for EXI tracing.
-    pub const EXI: &'static str = "SLIPPI_RUST_EXI";
-
-    /// Logs for `SlippiGameReporter`.
-    pub const GameReporter: &'static str = "SLIPPI_RUST_GAME_REPORTER";
+    pub const Slippi: &'static str = "SLIPPI_RUST_ONLINE";
```
Comment (NikhilNarayana) [inline dolphin/src/logger/mod.rs:50]:
```suggestion
    pub const SlippiOnline: &'static str = "SLIPPI_RUST_ONLINE";
```

either that or drop `Online` from the log name

File: dolphin/src/logger/mod.rs [hunk 4]
```diff
@@ -44,13 +44,10 @@ pub mod Log {
     /// Used for dumping logs from dependencies that we may need to inspect.
     /// This may also get logs that are not properly tagged! If that happens, you need
     /// to fix your logging calls. :)
-    pub const DEPENDENCIES: &'static str = "SLIPPI_RUST_DEPENDENCIES";
+    pub const Dependencies: &'static str = "SLIPPI_RUST_DEPENDENCIES";
 
     /// The default target for EXI tracing.
```
Comment (NikhilNarayana) [inline dolphin/src/logger/mod.rs:49]: update this comment

**Review by NikhilNarayana — CHANGES_REQUESTED** [review]

File: dolphin/src/logger/mod.rs (hunk 4 above)
Comment (ryanmcgrath) [reply dolphin/src/logger/mod.rs]: Done~

File: dolphin/src/logger/mod.rs (hunk 3 above)
Comment (ryanmcgrath) [reply dolphin/src/logger/mod.rs]: `Log::SlippiOnline` it is~

**Review by NikhilNarayana — APPROVED** [review]

**Pull request merged**